PL-48716 - FAQ addition (#6135)
harness/developer-hub: kb/platform/harness-platform-faqs.md

## Authentication

> ### As I migrate from First Generation (FG) to Next Generation (NG) and use Okta for SCIM, I've noticed our user groups have not migrated. Is there an additional step I need to take?

Reviewer: ```suggestion
### As I migrate from FirstGen to NextGen and use Okta for SCIM, I've noticed our user groups haven't migrated. Is there an additional step I need to take?
```

> Yes, to ensure that your user groups migrate correctly when moving from First Generation to Next Generation platforms and continue using Okta for SCIM, you will need to configure a new SCIM app specifically for the NG platform.

Reviewer: ```suggestion
Yes, to ensure that your user groups migrate correctly when moving from FirstGen to NextGen platforms and continue using Okta for SCIM, you must configure a new SCIM app specifically for the NextGen platform.
```

### Can we obtain the account's private key from the delegate without resorting to a PAT token?

No, it is not possible to obtain the account's private key from the delegate, even with a valid Personal Access Token (PAT). The delegate uses a secure token for encrypted communication with the Harness Manager. However, this communication framework does not provide a method for accessing the account's private key directly through the delegate. This limitation is by design to uphold security protocols. Delegates often operate with scoped permissions at the organization or project level, and allowing access to the account's private key via a delegate would pose a significant security risk. Therefore, regardless of the delegate's ability to communicate with the Harness Manager, access to the account's private key is strictly controlled and not available through delegate operations.

> ### Upon successful azure saml login for prod account Harness is redirecting to stage/dev account

Reviewer: ```suggestion
### Why is Harness redirecting to a stage/dev account upon a successful Azure SAML login for a prod account?
```

> Please check if entity id is set correctly in case if you are using multiple account, as if you are using vanity url then the entity id needs to be set accordingly while setting up saml in Harness as well as on azure side

Reviewer: ```suggestion
Ensure that the entity ID is set correctly if you are using multiple accounts. If you are using a vanity URL, the entity ID needs to be set accordingly when setting up SAML in Harness and on the Azure side.
```

## Delegates

> ### Is it possible to install multiple delegates of different types under the same name, and would this cause any issues with entities referring to the delegate?

Reviewer: ```suggestion
### Is it possible to install multiple delegates of different types under the same name, and would this cause any issues with entities referring to the delegate?
```

## Feature Flag

> ### After enabling the `PIE_MULTISELECT_AND_COMMA_IN_ALLOWED_VALUES` flag for my account, which allows for multiple selections within allowed values, do I have the option to choose when to enable multiple selections for an input, or is it automatically set to always allow multiple inputs?

Reviewer: ```suggestion
### After enabling the PIE_MULTISELECT_AND_COMMA_IN_ALLOWED_VALUES flag for my account, which allows for multiple selections within allowed values, do I have the option to choose when to enable multiple selections for an input, or is it automatically set to always allow multiple inputs?
```

Reviewer: @saifeemustafaqharness Using code ticks in headings throws the formatting off. Recommend removing.

However, alternative solutions exist for capturing such detailed information. For instance, App Dynamics supports the kind of sampling and detailed monitoring that can track method invocation timings, but do note that App Dynamics is not currently in use. Prometheus, while central for metrics management, lacks built-in support for detailed invocation timings without significant custom metric management.

For detailed tracing, including method invocation timing, integration with OpenTelemetry offers a promising approach. OpenTelemetry's tracing capabilities, combined with backend tools like Tempo (currently in prototyping), allow for capturing call graphs and execution times for specific trace IDs. OpenTelemetry also supports sampling strategies for high-latency requests, which can mitigate the overhead of capturing detailed execution data.

> If tracking the invocation timing of methods is necessary, utilizing the MetricService for recording durations (as Java Duration objects) or metrics (such as Unix time epochs) is recommended. An example provided includes capturing timestamps for each Delegate heartbeat event, showcasing a practical implementation for method invocation timing without overwhelming the metrics infrastructure.

Reviewer: ```suggestion
If tracking the invocation timing of methods is necessary, utilizing the MetricService for recording durations (as Java Duration objects) or metrics (such as Unix time epochs) is recommended. An example provided includes capturing timestamps for each delegate heartbeat event, showcasing a practical implementation for method invocation timing without overwhelming the metrics infrastructure.
```

Check the user group assigned to the user. If the user isn't assigned to any user group/role, they will not be able to view or access any relevant data.

## Pipelines

### How can I obtain the triggered build version value, trigger ID, or trigger URL during pipeline runtime when a pipeline is triggered by a Pull Request (PR)?

> To access information about the trigger during pipeline runtime, Harness provides specific expressions. For the username associated with the PR changes, you can use <+pipeline.triggeredBy.name>, which will give you the name of the user who initiated the PR. To get the ID of the trigger that fired the execution, the expression <+pipeline.triggeredBy.triggerIdentifier> can be utilized. However, it's important to note that Harness does not offer an expression to retrieve the URL to the trigger. Therefore, while you can easily access the username and trigger ID, obtaining the trigger URL directly in the pipeline runtime is not supported.

Reviewer: ```suggestion
Harness provides specific expressions to access information about the trigger during pipeline runtime. For the username associated with the PR changes, you can use `<+pipeline.triggeredBy.name>`, which will give you the name of the user who initiated the PR. To get the ID of the trigger that fired the execution, the expression `<+pipeline.triggeredBy.triggerIdentifier>` can be utilized. However, it's important to note that Harness does not offer an expression to retrieve the URL to the trigger. Therefore, while you can easily access the username and trigger ID, obtaining the trigger URL directly in the pipeline runtime is not supported.
```

Reviewer: @saifeemustafaqharness Adding code ticks to the expressions will fix your PR build.

## Settings

### Can I format the emails using the first letter of the first and last name in caps (e.g., FirstnameLastname@company.com) while adding them?

When adding emails, the specific formatting of uppercase letters, such as the first letter of the first and last name, does not impact the operation. This is because Harness performs a case-insensitive search for emails, converting all email addresses to lowercase during user creation operations. Therefore, regardless of how you format the capitalization in the email, it will be standardized to lowercase for processing.

> ### In the process of migrating SecretEngines from Classic to Next Generation (NG) platforms, we have encountered an issue with the secretEngineVersion set to 0 in a Vault example. However, when attempting to create this on NG, we face a constraint that secretEngineVersion cannot be 0. Is there a specific reason for this constraint, and should we simply increase the version number to 1?

Reviewer: ```suggestion
### In the process of migrating SecretEngines from Classic to NextGen platforms, we encountered an issue with the secretEngineVersion set to 0 in a Vault example. However, when attempting to create this on NextGen, we face a constraint that secretEngineVersion cannot be 0. Is there a specific reason for this constraint, and should we simply increase the version number to 1?
```

> Yes, the constraint you're encountering is due to validation rules that were introduced in the NG platform after its initial development. These validations are in place to ensure consistency and correctness across configurations. For the purpose of migrating SecretEngines from the Classic to the NG platform, it is recommended to adjust the secretEngineVersion from 0 to 1. The behaviour of versions 0 and 1 will be the same, with version 1 being the supported configuration in NG. This change is necessary to comply with NG's validation requirements and will facilitate a smooth transition of your SecretEngines to the new platform.

Reviewer: ```suggestion
Yes, the constraint you're encountering is due to validation rules that were introduced in the NextGen platform after its initial development. These validations are in place to ensure consistency and correctness across configurations. To migrate Secret Engines from the FirstGen to the NextGen platform, it is recommended to adjust the `secretEngineVersion` from 0 to 1. The behavior of versions 0 and 1 will be the same, with version 1 being the supported configuration in NextGen. This change is necessary to comply with NextGen's validation requirements and will facilitate a smooth transition of your Secret Engines to the new platform.
```
